src/components/swap/SwapBuyFiatButton.tsx

```diff
@@ -0,0 +1,144 @@
+
+export default function SwapBuyFiatButton() {
+  // Importing useWeb3React here is necessary for mocking different behavior in
+  // different tests in SwapBuyFiatButton.test.tsx
+
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const { useWeb3React } = require('@web3-react/core')
+  const { account } = useWeb3React()
+  const openFiatOnRampModal = useOpenModal(ApplicationModal.FIAT_ONRAMP)
+  const [buyFiatFlowCompleted, setBuyFiatFlowCompleted] = useBuyFiatFlowCompleted()
+  const [checkFiatRegionAvailability, setCheckFiatRegionAvailability] = useState(false)
+  const {
+    available: fiatOnrampAvailable,
+    availabilityChecked: fiatOnrampAvailabilityChecked,
+    loading: fiatOnrampAvailabilityLoading,
+  } = useFiatOnrampAvailability(checkFiatRegionAvailability)
+  const [buyFiatFlowState, setBuyFiatFlowState] = useState(BuyFiatFlowState.INACTIVE)
+  const [walletDrawerOpen, toggleWalletDrawer] = useWalletDrawer()
+
+  /* 
+  Once the buy fiat flow has been completed, reset BuyFiatFlowState and also save to 
+  local storage that user has completed this flow already.
+  */
+  const terminateBuyFiatFlow = useCallback(() => {
+    setBuyFiatFlowCompleted(true)
+    setBuyFiatFlowState(BuyFiatFlowState.INACTIVE)
+  }, [setBuyFiatFlowCompleted, setBuyFiatFlowState])
+
+  /*
+  Depending on the current state of the buy fiat flow the user is in (buyFiatFlowState),
+  the desired behavior of clicking the 'Buy' button is different:
+  1) Initially upon first click, need to check the availability of the feature in the user's 
+  region, and continue the flow.
+  2) If the feature is available in the user's region, need to connect a wallet, and continue
+  the flow.
+  3) If the feature is available and a wallet account is connected, show fiat on ramp modal.
+  4) If the feature is unavailable, show feature unavailable tooltip. 
+  */
+  const handleBuyCrypto = useCallback(() => {
+    if (!fiatOnrampAvailabilityChecked) {
+      setCheckFiatRegionAvailability(true)
+      setBuyFiatFlowState(BuyFiatFlowState.ACTIVE_POST_REGION_CHECK)
+    } else if (fiatOnrampAvailable && !account && !walletDrawerOpen) {
+      toggleWalletDrawer()
+      setBuyFiatFlowState(BuyFiatFlowState.ACTIVE_NEEDS_WALLET_CONNECTION)
+    } else if (fiatOnrampAvailable && account) {
+      openFiatOnRampModal()
+      terminateBuyFiatFlow()
+    } else if (!fiatOnrampAvailable) {
+      terminateBuyFiatFlow()
+    }
+  }, [
+    fiatOnrampAvailabilityChecked,
+    fiatOnrampAvailable,
+    account,
+    walletDrawerOpen,
+    toggleWalletDrawer,
+    openFiatOnRampModal,
+    terminateBuyFiatFlow,
+  ])
+
+  // Continue buy fiat flow automatically when requisite state changes have occured.
+  useEffect(() => {
+    if (
+      buyFiatFlowState === BuyFiatFlowState.ACTIVE_POST_REGION_CHECK ||
+      (account && buyFiatFlowState === BuyFiatFlowState.ACTIVE_NEEDS_WALLET_CONNECTION)
+    ) {
+      handleBuyCrypto()
```

Comment:
nit: I generally have nothing against `useEffect` for executing side-effects (it's one of the ways to do it React-way), but maybe... we can call `buyCrypto()` function directly from a click hnadler and don't really need state for it? Sounds like much easier way?